README.md
# MongodbLogger [![Build Status](https://secure.travis-ci.org/le0pard/mongodb_logger.png)](http://travis-ci.org/le0pard/mongodb_logger) [![Dependency Status](https://gemnasium.com/le0pard/mongodb_logger.png)](https://gemnasium.com/le0pard/mongodb_logger)

MongodbLogger is a alternative logger for Rails 3, which log all requests of you application into MongoDB database.
It:

* simple to integrate into existing Rails 3 application;
* allow to store all logs from web cluster into one scalable storage - MongoDB;
* flexible schema of MongoDB allow to store and search any information from logs;
* web panel allow filter logs, build graphs using MapReduce by information from logs;

## Rails 3 support

Please note the latest version is compatible with rails 3.1.x or newer.

For rails 3.0.x latest version 0.2.8.

## Installation

1. Add the following to your Gemfile then refresh your dependencies by executing "bundle install" (or just simple "bundle"):

        gem "mongodb_logger"

1. Add the following line to your ApplicationController:

        include MongodbLogger::Base

1. For use with Heroku you need to prevent the rails\_log\_stdout plugin from being added by Heroku:

        mkdir vendor/plugins/rails_log_stdout
        touch vendor/plugins/rails_log_stdout/.gitkeep

1. Add MongodbLogger settings to database.yml for each environment in which you want to use the MongodbLogger. The MongodbLogger will also
   look for a separate mongodb\_logger.yml or mongoid.yml (if you are using mongoid) before looking in database.yml.
   In the mongodb\_logger.yml and mongoid.yml case, the settings should be defined without the 'mongodb\_logger' subkey.

   database.yml:

        development:
         adapter: postgresql
         database: my_app_development
         username: postgres
         mongodb_logger:
           database: my_app               # required (the only required setting)
           capsize: <%= 10.megabytes %>   # default: 250MB
           host: localhost                # default: localhost
           port: 27017                    # default: 27017
           replica_set: true              # default: false - Adds retries for ConnectionFailure during voting for replica set master
           safe_insert: false             # default: false - Enable/Disable safe inserts (wait for insert to propagate to all nodes)
           application_name: my_app       # default: Rails.application
           disable_file_logging: false    # default: false - disable logging into filesystem (only in MongoDB)
           collection: some_name          # default: Rails.env + "_log" - name of MongoDB collection

   mongodb\_logger.yml:

        development:
          database: my_app
          capsize: <%= 10.megabytes %>
          host: localhost
          port: 27017
          replica_set: true

   Also you can use "url" parameter for setup connection to mongodb:

        development:
          url: mongodb://localhost:27017/my_app
          capsize: <%= 10.megabytes %>

   You can also use the "hosts" parameter v. "host" to specify a replset
        development:
          hosts:
            - mongo-host-01
            - mongo-host-02
            - mongo-host-03


1. For using with MongoDB Replica Set (more info you can read by this link [http://www.mongodb.org/display/DOCS/Replica+Sets](http://www.mongodb.org/display/DOCS/Replica+Sets)). In config set list of [host, port] in key "hosts":

        development:
          database: my_app
          capsize: <%= 10.megabytes %>
          host: localhost
          port: 27017
          hosts:
            - - 127.0.0.1
              - 27018
            - - 127.0.0.1
              - 27019

1. For assets pipeline you can generate all js/css file into folder by rake task:

        rake mongodb_logger:assets:compile[public/assets]

For capistrano possible compile assets by receipt. Add this to config/deploy.rb:

        require 'mongodb_logger/capistrano'
        set :mongodb_logger_assets_dir, "public/assets" # where to put mongodb assets
        after 'deploy:update_code', 'mongodb_logger:precompile'

Also you can serve assets from rails app. You need just mount it separately:

        mount MongodbLogger::Server.new, :at => "/mongodb", :as => :mongodb
        mount MongodbLogger::Assets.instance, :at => "/mongodb/assets", :as => :mongodb_assets # assets


## Usage

  After success instalation of gem, a new MongoDB document (record) will be created for each request on your application,
  by default will record the following information: Runtime, IP Address, Request Time, Controller, Method,
  Action, Params, Application Name and All messages sent to the logger. The structure of the MongoDB document looks like this:

      {
        'action'           : action_name,
        'application_name' : application_name (rails root),
        'controller'       : controller_name,
        'ip'               : ip_address,
        'messages'         : {
                               'info'  : [ ],
                               'debug' : [ ],
                               'error' : [ ],
                               'warn'  : [ ],
                               'fatal' : [ ]
                             },
        'params'           : { },
        'path'             : path,
        'request_time'     : date_of_request,
        'runtime'          : elapsed_execution_time_in_milliseconds,
        'url'              : full_url,
        'method'           : request method (GET, POST, OPTIONS),
        'is_exception'     : true only for exceptions (in other cases this field miss)
      }

  Beyond that, if you want to add extra information to the base of the document (let's say something like user\_id on every request that it's available),
  you can just call the Rails.logger.add\_metadata method on your logger like so (for example from a before\_filter):

      # make sure we're using the MongodbLogger in this environment
      if Rails.logger.respond_to?(:add_metadata)
       Rails.logger.add_metadata(:user_id => @current_user.id)
      end

## Callback on exceptions

  For send email or do something on exception you can add callback:

    MongodbLogger::Base.configure do |config|
      config.on_log_exception do |mongo_record|
        # do something with this data, for example - send email (better - by background job)
      end
    end

  In this callback send record without "\_id", because logger not wait for insert response from MongoDB.

## The Front End
  To setup web interface in you Rails application, first of all create autoload file in you Rails application

   File: you\_rails\_app/config/initializers/mongodb\_logger.rb (example)

        require 'mongodb_logger/server' # required
        # this secure you web interface by basic auth, but you can skip this, if you no need this
        MongodbLogger::Server.use Rack::Auth::Basic do |username, password|
            [username, password] == ['admin', 'password']
        end

   and just mount MongodbLogger::Server in rails routes:

   File: you\_rails\_app/config/routes.rb

        mount MongodbLogger::Server.new, :at => "/mongodb"

  Now you can see web interface by url "http://localhost:3000/mongodb"

  If you've installed MongodbLogger as a gem and want running the front end without Rails application, you can do it by this command:

      mongodb_logger_web config.yml

  where config.yml is config, similar to config of Rails apps, but without Rails.env. Example:

      database: app_logs_dev
      host: localhost
      port: 27017
      collection: development_log # set for see development logs

  parameter "collection" should be set, if your set custom for your Rails application or start this front end not for production
  enviroment (by default taken "production\_log" collection, in Rails application gem generate "#{Rails.env}\_log" collection,
  if it is not defined in config).

  It's a thin layer around rackup so it's configurable as well:

      mongodb_logger_web config.yml -p 8282

###  Passenger, Unicorn, Thin, etc.

  Using Passenger, Unicorn, Thin, etc? MongodbLogger ships with a `config.ru` you can use. See  guide:

  * Passenger Apache: <http://www.modrails.com/documentation/Users%20guide%20Apache.html#_deploying_a_rack_based_ruby_application>
  * Passenger Nginx: <http://www.modrails.com/documentation/Users%20guide%20Nginx.html#deploying_a_rack_app>
  * Unicorn: <http://unicorn.bogomips.org>
  * Thin: <http://code.macournoyer.com/thin/usage>

  Don't forget setup MONGODBLOGGERCONFIG env variable, which provide information about MongodbLogger config. Example starting with unicorn:

      MONGODBLOGGERCONFIG=examples/server_config.yml unicorn

##  Demo Application with MongodbLogger

  Demo: [http://demo-mongodb-logger.catware.org/](http://demo-mongodb-logger.catware.org/)

  Demo Sources: [https://github.com/le0pard/mongodb_logger_example_heroku](https://github.com/le0pard/mongodb_logger_example_heroku)


## Querying via the Rails console

  And now, for a couple quick examples on getting ahold of this log data...
  First, here's how to get a handle on the MongoDB from within a Rails console:

      >> db = Rails.logger.mongo_connection
      => #<Mongo::DB:0x007fdc7c65adc8 @name="monkey_logs_dev" ... >

      >> collection = db[Rails.logger.mongo_collection_name]
      => #<Mongo::Collection:0x007fdc7a4d12b0 @name="development_log" .. >

  Once you've got the collection, you can find all requests for a specific user (with id):

      >> cursor = collection.find(:user_id => '12355')
      => #<Mongo::Cursor:0x1031a3e30 ... >
      >> cursor.count
      => 5

  Find all requests that took more that one second to complete:

      >> collection.find({:runtime => {'$gt' => 1000}}).count
      => 3

  Find all order#show requests with a particular order id (id=order\_id):

      >> collection.find({"controller" => "order", "action"=> "show", "params.id" => order_id})

  Find all requests with an exception that contains "RoutingError" in the message or stack trace:

      >> collection.find({"messages.error" => /RoutingError/})

  Find all requests with errors:

      >> collection.find({"is_exception" => true})

  Find all requests with a request\_date greater than '11/18/2010 22:59:52 GMT'

      >> collection.find({:request_time => {'$gt' => Time.utc(2010, 11, 18, 22, 59, 52)}})



Copyright (c) 2009-2012 Phil Burrows, CustomInk (based on https://github.com/customink/central_logger) and Leopard released under the MIT license


CHANGELOG.md
## v0.3.4

* added support for mongo replsets via "hosts" parameter

## v0.3.3

* Mount assets separately or compile it by rake task 
* Capistrano recipe for compiling assets

## v0.3.2

* Initialize a connection to MongoDB using the MongoDB URI spec
* Fix logger.info/warn/error for Rails 3.2.x

## v0.3.1

* Fix broken images in if using assets pipeline

## v0.3.0

* Added assets pipeline support

## v0.2.8

* Added callback on exception
* Delete deprecate warnings

## v0.2.7

* Fixed error trace for ruby 1.9.2
* Update gems dependency

## v0.2.6

* Replica Set set in config by key 'hosts'
* Adding jruby support

## v0.2.5

* Build graphs for analytics page
* Fix month for analytics

## v0.2.4

* Move on logs list by using arrows on keybords

## v0.2.3

* Fix storage log with attachments (save it for search by "original\_filename" and "content\_type")
* Fix tests for CI

## v0.2.2

* Fix set custom collection in config and add tests on it 

## v0.2.1

* Add show on 'more info' page meta data from log

## v0.2.0

* Public release

test/unit/mongodb_logger_replica_test.rb
require 'test_helper'
require 'mongodb_logger/logger'

# HOWTO run this test:
# before start this test, do this in console:
# mkdir -p /tmp/data1
# mkdir -p /tmp/data2
# mkdir -p /tmp/data3
# mongod --replSet foo --port 27018 --dbpath /tmp/data1
# mongod --replSet foo --port 27019 --dbpath /tmp/data2
# mongod --replSet foo --port 27020 --dbpath /tmp/data3
# mongo localhost:27018
# In mongo console:
# config = {_id: 'foo', members: [
# {_id: 0, host: 'localhost:27018'},
# {_id: 1, host: 'localhost:27019'},
# {_id: 2, host: 'localhost:27020', arbiterOnly: true}]
# }
# 
# rs.initiate(config);
# You should see such output:
# {
#         "info" : "Config now saved locally.  Should come online in about a minute.",
#         "ok" : 1
# }


class MongodbLogger::MongodbLoggerReplicaTest < Test::Unit::TestCase
  extend LogMacros

  context "A MongodbLogger::MongoLogger connecting to a replica set" do
    setup do
      FileUtils.cp(File.join(SAMPLE_CONFIG_DIR, REPLICA_SET_CONFIG),  File.join(CONFIG_DIR, DEFAULT_CONFIG))
      MongodbLogger::Logger.any_instance.stubs(:internal_initialize).returns(nil)
      MongodbLogger::Logger.any_instance.stubs(:disable_file_logging?).returns(false)
      @mongodb_logger = MongodbLogger::Logger.new
      @mongodb_logger.send(:configure)
      @mongodb_logger.send(:connect)
      common_setup
      @collection.drop
    end

    should "derive from Mongo::ReplSetConnection" do
      assert_equal Mongo::ReplSetConnection, @mongodb_logger.mongo_connection_type
    end

    should "force replica_set parameter to be true" do
      assert @mongodb_logger.db_configuration['replica_set']
    end

    teardown do
      file = File.join(CONFIG_DIR, DEFAULT_CONFIG)
      File.delete(file) if File.exist?(file)
    end
  end
end


lib/mongodb_logger/server_config.rb
require 'mongo'
require 'erb'
require 'active_support'
require 'active_support/core_ext'

# TODO: Dry this class with logger class
module MongodbLogger
  class ServerConfig

    DEFAULT_COLLECTION_SIZE = 250.megabytes

    class << self

      def set_config(config_path)
        if File.file?(config_path)
          config_file = File.new(config_path)
          config = YAML.load(ERB.new(config_file.read).result)
        else
          raise "Config file not found"
        end

        # default to localhost
        @mongo_cfg = {
          'host' => 'localhost',
          'port' => 27017
        }.merge(config)

        # specify a collection or default to name based on environment
        @mongo_cfg["collection"] ||= "#{ Rails.env }_log"

        # configure mongo based hosts (replset) or host (standalone)
        if @mongo_cfg['hosts']
          replset  = Mongo::MongoReplicaSetClient.new(@mongo_cfg['hosts'].map { |h| "#{h}:#{@mongo_cfg['port']}" },
                                                     :refresh_mode => false,
                                                     :read         => :secondary_preferred)
          @mongodb = replset.db(@mongo_cfg["database"])
        else
          # if in dev or staging mode, it's ok to connect to a mongo slave server
          slave_ok  = Rails.env.development? || Rails.env.staging?
          @mongodb  = Mongo::MongoClient.new(@mongo_cfg["host"], @mongo_cfg["port"], :slave_ok => slave_ok).db(@mongo_cfg["database"])
        end

        # see if we need to authenticate
        @mongodb.authenticate(MONGO_CONFIG["username"], MONGO_CONFIG["password"]) unless MONGO_CONFIG["username"].nil?

        set_collection
      end

      def set_config_for_testing(config_path)
        set_config(config_path)
        create_collection unless @mongodb.collection_names.include?(@mongo_cfg["collection"])
        set_collection
      end

      def create_collection
        capsize = DEFAULT_COLLECTION_SIZE
        capsize = @mongo_cfg['capsize'].to_i if @mongo_cfg['capsize']
        @mongodb.create_collection(@mongo_cfg["collection"],
                                            {:capped => true, :size => capsize})
      end


      def set_collection
        @collection = @mongodb[@mongo_cfg["collection"]]
      end

      def get_config
        @mongo_cfg
      end

      def authenticated?
        @authenticated
      end

      def collection_name
        @mongo_cfg["collection"]
      end

      def db
        @mongodb
      end

      def collection
        @collection
      end
    end
  end
end


lib/mongodb_logger/logger.rb
require 'erb'
require 'uri'
require 'mongo'
require 'active_support'
require 'active_support/core_ext'
require 'active_support/core_ext/logger'
require 'action_dispatch/http/upload'
require 'mongodb_logger/replica_set_helper'

module MongodbLogger
  class Logger < ActiveSupport::BufferedLogger
    include ReplicaSetHelper

    DEFAULT_COLLECTION_SIZE = 250.megabytes
    # Looks for configuration files in this order
    CONFIGURATION_FILES = ['mongo.yml', 'mongodb_logger.yml', 'mongoid.yml', 'database.yml']
    LOG_LEVEL_SYM = [:debug, :info, :warn, :error, :fatal, :unknown]

    attr_reader :db_configuration, :mongo_connection, :mongo_collection_name, :mongo_collection, :mongo_connection_type

    def initialize(options={})
      path = options[:path] || File.join(Rails.root, "log/#{Rails.env}.log")
      @level = options[:level] || DEBUG
      internal_initialize
    rescue => e
      # should use a config block for this
      Rails.env.production? ? (raise e) : (puts 'MongodbLogger WARNING: Using BufferedLogger due to exception: ' + e.message)
    ensure
      if disable_file_logging?
        @log            = ::Logger.new(STDOUT)
        @log.level      = @level
      else
        super(path, @level)
      end
    end

    def add_metadata(options={})
      options.each do |key, value|
        unless [:messages, :request_time, :ip, :runtime, :application_name, :is_exception, :params, :session, :method].include?(key.to_sym)
          @mongo_record[key] = value
        else
          raise ArgumentError, ":#{key} is a reserved key for the mongodb logger. Please choose a different key"
        end
      end if @mongo_record
    end

    def add(severity, message = nil, progname = nil, &block)
      $stdout.puts(message) if ENV['HEROKU_RACK'] # log in stdout on Heroku
      if @level && @level <= severity && message.present? && @mongo_record.present?
        # do not modify the original message used by the buffered logger
        msg = logging_colorized? ? message.to_s.gsub(/(\e(\[([\d;]*[mz]?))?)?/, '').strip : message
        @mongo_record[:messages][LOG_LEVEL_SYM[severity]] << msg
      end
      # may modify the original message
      disable_file_logging? ? message : (@level ? super : message)
    end

    # Drop the capped_collection and recreate it
    def reset_collection
      if @mongo_connection && @mongo_collection
        @mongo_collection.drop
        create_collection
      end
    end

    def mongoize(options={})
      @mongo_record = options.merge({
        :messages => Hash.new { |hash, key| hash[key] = Array.new },
        :request_time => Time.now.getutc,
        :application_name => @application_name
      })

      runtime = Benchmark.measure{ yield }.real if block_given?
    rescue Exception => e
      add(3, e.message.to_s + '\n' + e.backtrace.join('\n'))
      # log exceptions
      @mongo_record[:is_exception] = true
      # Reraise the exception for anyone else who cares
      raise e
    ensure
      # In case of exception, make sure runtime is set
      @mongo_record[:runtime] = ((runtime ||= 0) * 1000).ceil
      # error callback
      Base.on_log_exception(@mongo_record) if @mongo_record[:is_exception]
      begin
        @insert_block.call
      rescue
        begin
          # try to nice serialize record
          nice_serialize @mongo_record
          @insert_block.call
        rescue
          # do extra work to inpect (and flatten)
          force_serialize @mongo_record
          @insert_block.call rescue nil
        end
      end
    end

    def authenticated?
      @authenticated
    end

    private
      # facilitate testing
      def internal_initialize
        configure
        connect
        check_for_collection
      end

      def disable_file_logging?
        @db_configuration.fetch('disable_file_logging', false)
      end

      def configure
        default_capsize = DEFAULT_COLLECTION_SIZE
        @authenticated = false
        @db_configuration = {
          'host' => 'localhost',
          'port' => 27017,
          'capsize' => default_capsize}.merge(resolve_config)
        @mongo_collection_name = @db_configuration['collection'] || "#{Rails.env}_log"
        @application_name = resolve_application_name
        @safe_insert = @db_configuration['safe_insert'] || false

        @insert_block = @db_configuration.has_key?('replica_set') && @db_configuration['replica_set'] ?
          lambda { rescue_connection_failure{ insert_log_record(@safe_insert) } } :
          lambda { insert_log_record }
      end

      def resolve_application_name
        if @db_configuration.has_key?('application_name')
          @db_configuration['application_name']
        else
          Rails.application.class.to_s.split('::').first
        end
      end

      def resolve_config
        config = {}
        CONFIGURATION_FILES.each do |filename|
          config_file = Rails.root.join('config', filename)
          if config_file.file?
            config = YAML.load(ERB.new(config_file.read).result)[Rails.env]
            config = config['mongodb_logger'] if config && config.has_key?('mongodb_logger')
            break unless config.blank?
          end
        end
        config
      end

      def mongo_connection_object
        if @db_configuration['hosts']
          conn = Mongo::ReplSetConnection.new(*(@db_configuration['hosts'] <<
            {:connect => true, :pool_timeout => 6}))
          @db_configuration['replica_set'] = true
        elsif @db_configuration['url']
          conn = Mongo::Connection.from_uri(@db_configuration['url'])
        else
          conn = Mongo::Connection.new(@db_configuration['host'],
                                       @db_configuration['port'],
                                       :connect => true,
                                       :pool_timeout => 6)
        end
        @mongo_connection_type = conn.class
        conn
      end

      def connect
        if @db_configuration['url']
          uri = URI.parse(@db_configuration['url'])
          @mongo_connection ||= mongo_connection_object.db(uri.path.gsub(/^\//, ''))
          @authenticated = true
        else
          @mongo_connection ||= mongo_connection_object.db(@db_configuration['database'])
          if @db_configuration['username'] && @db_configuration['password']
            # the driver stores credentials in case reconnection is required
            @authenticated = @mongo_connection.authenticate(@db_configuration['username'],
                                                          @db_configuration['password'])
          end
        end
      end

      def create_collection
        @mongo_connection.create_collection(@mongo_collection_name,
                                            {:capped => true, :size => @db_configuration['capsize'].to_i})
      end

      def check_for_collection
        # setup the capped collection if it doesn't already exist
        create_collection unless @mongo_connection.collection_names.include?(@mongo_collection_name)
        @mongo_collection = @mongo_connection[@mongo_collection_name]
      end

      def insert_log_record(safe = false)
        @mongo_collection.insert(@mongo_record, :safe => safe)
      end

      def logging_colorized?
        # Cache it since these ActiveRecord attributes are assigned after logger initialization occurs in Rails boot
        @colorized ||= Object.const_defined?(:ActiveRecord) && ActiveRecord::LogSubscriber.colorize_logging
      end

      # try to serialyze data by each key and found invalid object
      def nice_serialize(rec)
        if msgs = rec[:messages]
          msgs.each do |i, j|
            msgs[i] = nice_serialize_object(j)
          end
        end
        if pms = rec[:params]
          pms.each do |i, j|
            pms[i] = nice_serialize_object(j)
          end
        end
      end

      def nice_serialize_object(data)
        case data
          when NilClass, String, Fixnum, Bignum, Float, TrueClass, FalseClass, Time, Regexp, Symbol
            data
          when Hash
            hvalues = Hash.new
            data.each{|k,v| hvalues[k] = nice_serialize_object(v) }
            hvalues
          when Array
            data.map{|v| nice_serialize_object(v) }
          when ActionDispatch::Http::UploadedFile, Rack::Test::UploadedFile # uploaded files
            hvalues = {
              :original_filename => data.original_filename,
              :content_type => data.content_type
            }
          else
            data.inspect
        end
      end

      # force the data in the db by inspecting each top level array and hash element
      # this will flatten other hashes and arrays
      def force_serialize(rec)
        if msgs = rec[:messages]
          msgs.each { |i, j| msgs[i] = j.inspect }
        end
        if pms = rec[:params]
          pms.each { |i, j| pms[i] = j.inspect }
        end
      end
  end
end
